File: pricing/02-billing-history/billing_history.ts
/**
 * Billing history viewer for HarchOS GPU workloads (TypeScript).
 *
 * Retrieves and displays billing records from the HarchOS platform
 * using the JS SDK. Shows filtering by status and date range, and
 * computes spending summaries with overdue alerts.
 *
 * Usage:
 *   npx ts-node billing_history.ts
 *   npx ts-node billing_history.ts --status overdue
 *   HARCHOS_API_KEY=hsk_... npx ts-node billing_history.ts
 */

import { HarchOSClient } from "harchos";
import type { BillingRecord, ListBillingRecordsParams } from "harchos";

// ---------------------------------------------------------------------------
// Simulated billing data
// ---------------------------------------------------------------------------

interface SimulatedRecord {
  id: string;
  userId: string;
  workloadId: string;
  hubId: string;
  gpuHours: number;
  cpuCoreHours: number;
  memoryGbHours: number;
  storageGbMonths: number;
  totalCost: number;
  currency: string;
  status: "open" | "closed" | "paid" | "overdue";
  periodStart: string;
  periodEnd: string;
}

const SIMULATED_RECORDS: SimulatedRecord[] = [
  {
    id: "bill-001",
    userId: "user-alice",
    workloadId: "wkld-training-resnet",
    hubId: "hub-ouarzazate",
    gpuHours: 192,
    cpuCoreHours: 768,
    memoryGbHours: 3072,
    storageGbMonths: 50,
    totalCost: 403.2,
    currency: "USD",
    status: "paid",
    periodStart: "2025-01-01T00:00:00Z",
    periodEnd: "2025-01-31T23:59:59Z",
  },
  {
    id: "bill-002",
    userId: "user-alice",
    workloadId: "wkld-llama-inference",
    hubId: "hub-dakhla",
    gpuHours: 720,
    cpuCoreHours: 1440,
    memoryGbHours: 5760,
    storageGbMonths: 200,
    totalCost: 1008,
    currency: "USD",
    status: "paid",
    periodStart: "2025-01-01T00:00:00Z",
    periodEnd: "2025-01-31T23:59:59Z",
  },
  {
    id: "bill-003",
    userId: "user-bob",
    workloadId: "wkld-finetune-mistral",
    hubId: "hub-benguerir",
    gpuHours: 336,
    cpuCoreHours: 1344,
    memoryGbHours: 5376,
    storageGbMonths: 100,
    totalCost: 789.6,
    currency: "USD",
    status: "open",
    periodStart: "2025-02-01T00:00:00Z",
    periodEnd: "2025-02-28T23:59:59Z",
  },
  {
    id: "bill-004",
    userId: "user-carol",
    workloadId: "wkld-rag-pipeline",
    hubId: "hub-tanger",
    gpuHours: 48,
    cpuCoreHours: 192,
    memoryGbHours: 768,
    storageGbMonths: 500,
    totalCost: 86.4,
    currency: "USD",
    status: "overdue",
    periodStart: "2024-12-01T00:00:00Z",
    periodEnd: "2024-12-31T23:59:59Z",
  },
];

// ---------------------------------------------------------------------------
// Helpers
// ---------------------------------------------------------------------------

const STATUS_ICONS: Record<string, string> = {
  paid: "✅",
  open: "🟡",
  closed: "🔵",
  overdue: "🔴",
};

function printRecords(records: SimulatedRecord[]): void {
  if (!records.length) {
    console.log("\n  No billing records found.");
    return;
  }

  console.log(
    `\n  ${"ID".padEnd(12)} | ${"Workload".padEnd(28)} | ${"Hub".padEnd(18)} | ` +
    `${"GPU hrs".padStart(8)} | ${"Total".padStart(10)} | ${"Currency".padEnd(8)} | Status`
  );
  console.log("  " + "-".repeat(110));

  for (const r of records) {
    const icon = STATUS_ICONS[r.status] || "⚪";
    console.log(
      `  ${r.id.padEnd(12)} | ${(r.workloadId || "N/A").padEnd(28)} | ` +
      `${(r.hubId || "N/A").padEnd(18)} | ` +
      `${r.gpuHours.toFixed(1).padStart(8)} | ` +
      `${r.totalCost.toFixed(2).padStart(10)} | ${r.currency.padEnd(8)} | ` +
      `${icon} ${r.status}`
    );
  }
}

interface CurrencySummary {
  total: number;
  byStatus: Record<string, number>;
  totalGpuHours: number;
  recordCount: number;
}

function computeSummary(records: SimulatedRecord[]): Record<string, CurrencySummary> {
  const summary: Record<string, CurrencySummary> = {};

  for (const r of records) {
    if (!summary[r.currency]) {
      summary[r.currency] = { total: 0, byStatus: {}, totalGpuHours: 0, recordCount: 0 };
    }
    const s = summary[r.currency];
    s.total += r.totalCost;
    s.totalGpuHours += r.gpuHours;
    s.recordCount += 1;
    s.byStatus[r.status] = (s.byStatus[r.status] || 0) + r.totalCost;
  }

  return summary;
}

function printSummary(summary: Record<string, CurrencySummary>): void {
  console.log("\n  📊 Spending Summary");
  console.log("  " + "=".repeat(50));

  for (const [currency, data] of Object.entries(summary)) {
    console.log(`\n  Currency: ${currency}`);
    console.log(`    Total spent:      ${data.total.toFixed(2).padStart(10)} ${currency}`);
    console.log(`    Total GPU hours:  ${data.totalGpuHours.toFixed(1).padStart(10)} hrs`);
    console.log(`    Record count:     ${data.recordCount.toString().padStart(10)}`);

    console.log("    By status:");
    for (const [status, amount] of Object.entries(data.byStatus)) {
      const icon = STATUS_ICONS[status] || "⚪";
      console.log(`      ${icon} ${status.padEnd(10)}: ${amount.toFixed(2).padStart(10)} ${currency}`);
    }

    const overdue = data.byStatus["overdue"] || 0;
    if (overdue > 0) {
      console.log(`\n    ⚠️  OVERDUE: ${overdue.toFixed(2)} ${currency} requires immediate attention!`);
    }
  }
}

// ---------------------------------------------------------------------------
// Main
// ---------------------------------------------------------------------------

async function main(): Promise<void> {
  const args = process.argv.slice(2);
  const statusFilter = getArg(args, "--status", "") as ListBillingRecordsParams["status"] | "";

  console.log("=".repeat(60));
  console.log("HarchOS Billing History (TypeScript)");
  console.log("=".repeat(60));

  let records: SimulatedRecord[];

  // Try SDK, fall back to simulated data
  const apiKey = process.env.HARCHOS_API_KEY || "";
  if (apiKey) {
    try {
      const client = new HarchOSClient({ apiKey });
      const params: ListBillingRecordsParams = {};
      if (statusFilter) params.status = statusFilter;

      const sdkRecords = await client.pricing.listBillingRecords(params);
      records = sdkRecords.map((r) => ({
        id: r.id,
        userId: r.userId,
        workloadId: r.workloadId || "N/A",
        hubId: r.hubId || "N/A",
        gpuHours: r.gpuHours,
        cpuCoreHours: r.cpuCoreHours,
        memoryGbHours: r.memoryGbHours,
        storageGbMonths: r.storageGbMonths,
        totalCost: r.totalCost,
        currency: r.currency,
        status: r.status,
        periodStart: r.periodStart,
        periodEnd: r.periodEnd,
      }));
    } catch (err) {
      console.error(`[ERROR] SDK call failed: ${err}`);
      console.log("[WARN] Falling back to simulated data");
      records = getSimulatedRecords(statusFilter);
    }
  } else {
    records = getSimulatedRecords(statusFilter);
  }

  printRecords(records);

  if (records.length) {
    const summary = computeSummary(records);
    printSummary(summary);
  }

  console.log();
}

function getSimulatedRecords(statusFilter: string): SimulatedRecord[] {
  if (statusFilter) {
    return SIMULATED_RECORDS.filter((r) => r.status === statusFilter);
  }
  return SIMULATED_RECORDS;
}

function getArg(args: string[], flag: string, defaultValue: string): string {
  const idx = args.indexOf(flag);
  if (idx !== -1 && idx + 1 < args.length) return args[idx + 1];
  return defaultValue;
}

main().catch((err) => {
  console.error("Fatal error:", err);
  process.exit(1);
});

File: pricing/02-billing-history/billing_history.py
#!/usr/bin/env python3
"""Billing history viewer for HarchOS GPU workloads.

Retrieves and displays billing records from the HarchOS platform using
the Python SDK. Shows how to filter records by status, date range, and
compute spending summaries.

Features:
- List billing records with filters
- Filter by status (open, closed, paid, overdue)
- Date range filtering
- Spending summary with totals by status and currency
- Overdue record alerts

Usage:
    python billing_history.py
    python billing_history.py --status overdue
    python billing_history.py --last-30-days
    HARCHOS_API_KEY=hsk_... python billing_history.py
"""

import argparse
import json
import os
from datetime import datetime, timedelta, timezone
from typing import Dict, List, Optional

try:
    from harchos import HarchOS
    HARCHOS_AVAILABLE = True
except ImportError:
    HARCHOS_AVAILABLE = False
    print("[WARN] harchos SDK not installed — using simulated billing data")


# ---------------------------------------------------------------------------
# Simulated billing data (for local testing)
# ---------------------------------------------------------------------------

SIMULATED_RECORDS = [
    {
        "id": "bill-001",
        "user_id": "user-alice",
        "workload_id": "wkld-training-resnet",
        "hub_id": "hub-ouarzazate",
        "gpu_hours": 192.0,
        "cpu_core_hours": 768.0,
        "memory_gb_hours": 3072.0,
        "storage_gb_months": 50.0,
        "total_cost": 403.20,
        "currency": "USD",
        "status": "paid",
        "period_start": "2025-01-01T00:00:00Z",
        "period_end": "2025-01-31T23:59:59Z",
    },
    {
        "id": "bill-002",
        "user_id": "user-alice",
        "workload_id": "wkld-llama-inference",
        "hub_id": "hub-dakhla",
        "gpu_hours": 720.0,
        "cpu_core_hours": 1440.0,
        "memory_gb_hours": 5760.0,
        "storage_gb_months": 200.0,
        "total_cost": 1008.00,
        "currency": "USD",
        "status": "paid",
        "period_start": "2025-01-01T00:00:00Z",
        "period_end": "2025-01-31T23:59:59Z",
    },
    {
        "id": "bill-003",
        "user_id": "user-bob",
        "workload_id": "wkld-finetune-mistral",
        "hub_id": "hub-benguerir",
        "gpu_hours": 336.0,
        "cpu_core_hours": 1344.0,
        "memory_gb_hours": 5376.0,
        "storage_gb_months": 100.0,
        "total_cost": 789.60,
        "currency": "USD",
        "status": "open",
        "period_start": "2025-02-01T00:00:00Z",
        "period_end": "2025-02-28T23:59:59Z",
    },
    {
        "id": "bill-004",
        "user_id": "user-carol",
        "workload_id": "wkld-rag-pipeline",
        "hub_id": "hub-tanger",
        "gpu_hours": 48.0,
        "cpu_core_hours": 192.0,
        "memory_gb_hours": 768.0,
        "storage_gb_months": 500.0,
        "total_cost": 86.40,
        "currency": "USD",
        "status": "overdue",
        "period_start": "2024-12-01T00:00:00Z",
        "period_end": "2024-12-31T23:59:59Z",
    },
    {
        "id": "bill-005",
        "user_id": "user-alice",
        "workload_id": "wkld-etl-pipeline",
        "hub_id": "hub-ouarzazate",
        "gpu_hours": 0.0,
        "cpu_core_hours": 2000.0,
        "memory_gb_hours": 4000.0,
        "storage_gb_months": 1000.0,
        "total_cost": 210.00,
        "currency": "MAD",
        "status": "closed",
        "period_start": "2025-01-15T00:00:00Z",
        "period_end": "2025-02-14T23:59:59Z",
    },
]


class SimulatedBillingClient:
    """Simulated billing client for local testing."""

    def list_billing_records(self, **filters):
        """Return simulated billing records, applying filters."""
        records = SIMULATED_RECORDS
        if filters.get("status"):
            records = [r for r in records if r["status"] == filters["status"]]
        return records


# ---------------------------------------------------------------------------
# Billing History Viewer
# ---------------------------------------------------------------------------

class BillingHistoryViewer:
    """Retrieves and analyzes billing history from HarchOS."""

    def __init__(self):
        if HARCHOS_AVAILABLE:
            self.client = HarchOS(api_key=os.environ.get("HARCHOS_API_KEY", ""))
        else:
            self.client = SimulatedBillingClient()

    def get_records(
        self,
        status: Optional[str] = None,
        period_start: Optional[datetime] = None,
        period_end: Optional[datetime] = None,
        limit: Optional[int] = None,
    ) -> List[dict]:
        """Fetch billing records with optional filters."""
        try:
            if HARCHOS_AVAILABLE:
                records = self.client.pricing.list_billing_records(
                    status=status,
                    period_start=period_start,
                    period_end=period_end,
                    limit=limit,
                )
                # Convert SDK models to dicts for display
                return [
                    {
                        "id": r.id,
                        "user_id": r.user_id,
                        "workload_id": r.workload_id,
                        "hub_id": r.hub_id,
                        "gpu_hours": r.gpu_hours,
                        "cpu_core_hours": r.cpu_core_hours,
                        "memory_gb_hours": r.memory_gb_hours,
                        "storage_gb_months": r.storage_gb_months,
                        "total_cost": r.total_cost,
                        "currency": r.currency,
                        "status": r.status,
                        "period_start": str(r.period_start),
                        "period_end": str(r.period_end),
                    }
                    for r in records
                ]
            else:
                return self.client.list_billing_records(
                    status=status,
                    period_start=period_start,
                    period_end=period_end,
                )
        except Exception as e:
            print(f"[ERROR] Failed to retrieve billing records: {e}")
            return []

    def print_records(self, records: List[dict]):
        """Pretty-print billing records in a table format."""
        if not records:
            print("\n  No billing records found.")
            return

        print(f"\n  {'ID':12s} | {'Workload':28s} | {'Hub':18s} | {'GPU hrs':>8s} | "
              f"{'Total':>10s} | {'Currency':8s} | {'Status':8s}")
        print("  " + "-" * 110)

        for r in records:
            status_icon = {
                "paid": "✅",
                "open": "🟡",
                "closed": "🔵",
                "overdue": "🔴",
            }.get(r["status"], "⚪")

            print(f"  {r['id']:12s} | {(r.get('workload_id') or 'N/A'):28s} | "
                  f"{(r.get('hub_id') or 'N/A'):18s} | "
                  f"{r['gpu_hours']:8.1f} | "
                  f"{r['total_cost']:10.2f} | {r['currency']:8s} | "
                  f"{status_icon} {r['status']}")

    def compute_summary(self, records: List[dict]) -> dict:
        """Compute spending summary from billing records."""
        summary: Dict[str, dict] = {}

        for r in records:
            currency = r["currency"]
            if currency not in summary:
                summary[currency] = {
                    "total": 0.0,
                    "by_status": {},
                    "total_gpu_hours": 0.0,
                    "record_count": 0,
                }

            summary[currency]["total"] += r["total_cost"]
            summary[currency]["total_gpu_hours"] += r["gpu_hours"]
            summary[currency]["record_count"] += 1

            status = r["status"]
            if status not in summary[currency]["by_status"]:
                summary[currency]["by_status"][status] = 0.0
            summary[currency]["by_status"][status] += r["total_cost"]

        return summary

    def print_summary(self, summary: dict):
        """Pretty-print spending summary."""
        print("\n  📊 Spending Summary")
        print("  " + "=" * 50)

        for currency, data in summary.items():
            print(f"\n  Currency: {currency}")
            print(f"    Total spent:      {data['total']:>10.2f} {currency}")
            print(f"    Total GPU hours:  {data['total_gpu_hours']:>10.1f} hrs")
            print(f"    Record count:     {data['record_count']:>10d}")

            print(f"    By status:")
            for status, amount in data["by_status"].items():
                icon = {"paid": "✅", "open": "🟡", "closed": "🔵", "overdue": "🔴"}.get(status, "⚪")
                print(f"      {icon} {status:10s}: {amount:>10.2f} {currency}")

            # Alert on overdue
            overdue = data["by_status"].get("overdue", 0)
            if overdue > 0:
                print(f"\n    ⚠️  OVERDUE: {overdue:.2f} {currency} requires immediate attention!")


# ---------------------------------------------------------------------------
# Main
# ---------------------------------------------------------------------------

def main():
    parser = argparse.ArgumentParser(description="HarchOS Billing History")
    parser.add_argument("--status", type=str, default=None,
                        choices=["open", "closed", "paid", "overdue"],
                        help="Filter by billing status")
    parser.add_argument("--last-30-days", action="store_true",
                        help="Show records from the last 30 days")
    parser.add_argument("--limit", type=int, default=None,
                        help="Maximum number of records")
    args = parser.parse_args()

    viewer = BillingHistoryViewer()

    # Parse date range
    period_start = None
    period_end = None
    if args.last_30_days:
        period_end = datetime.now(timezone.utc)
        period_start = period_end - timedelta(days=30)

    print("=" * 60)
    print("HarchOS Billing History")
    print("=" * 60)

    # Fetch records
    records = viewer.get_records(
        status=args.status,
        period_start=period_start,
        period_end=period_end,
        limit=args.limit,
    )

    # Display records
    viewer.print_records(records)

    # Compute and display summary
    if records:
        summary = viewer.compute_summary(records)
        viewer.print_summary(summary)

    print()


if __name__ == "__main__":
    main()
